### PR TITLE
Try menu appear animation & start work on handbook

### PR DESCRIPTION
**Pull request by jasmussen**
This PR aims to start work on #8029, by creating an animation for use with popovers, and by starting documentation for hwo to best use animation.

It:

- Adds a "scale and appear" animation to the block toolbar, the more menu, and the block library
- It adds a very very early draft of an animation document, which includes an early inventory of current animations in use.

GIF:

popover appear

### CODE REVIEW

File: packages/components/src/popover/style.scss
Comment (gziolo) [reply packages/components/src/popover/style.scss]: Or generalized and controlled by the value passed through prop translated into a corresponding class name.

**Comment by jasmussen** [comment]
> I don't see this, did I miss something?

Ack, looks like I forgot to `git add the new MD file, and now I'm relocating home! I'll be sure to commit it tomorrow, sorry about that.

File: packages/components/src/popover/style.scss
Comment (youknowriad) [reply packages/components/src/popover/style.scss]: Maybe instead, this would be a HoC/renderprop/hook adding the correct className based on the "position" (top left,...)

**Comment by youknowriad** [comment]
I'm looking forward adding some subtle animations to Gutenberg. Thanks for starting this Joen.

File: packages/components/src/popover/style.scss
Comment (gziolo) [reply packages/components/src/popover/style.scss]: Yep, something like that :)

**Comment by folletto** [comment]
On principle: good move, great to get this started, nice effect. 🌟 

I feel this is the kind of thing I'd push live early right after a release so people can feel it. 

Aside: I consider 200ms the upper limit for these animations. So from a theoretical perspective I think we're good.

Have we checked the performance on mobile?

**Comment by melchoyce** [comment]
FWIW, the animation feels a little slow to me.

**Comment by kjellr** [comment]
Agreed with @melchoyce and @folletto on the speed. I tried speeding it up to `0.01s` and it feels pretty good to me. (It could _maybe_ even be a little faster than that, but `0.05s` seemed a little too quick). 

I did see this error, which doesn't happen to me in `master`:

inserter

The inserter there pops up, even though there isn't enough space. 🤔 

It also seemed weird that the Content Structure and Block Navigation menus didn't have this animation... but I'm guessing those use a different component or something? 

animation

**Comment by mapk** [comment]
I experienced the bug that Kjell mentioned as well. And also agree on speeding up the animation a tad. It's looking so beautiful... and FUN!

**Comment by jasmussen** [comment]
THANK you all for the wonderful feedback! I heartily agree, the animation should be faster, here it is in .1s:

appear

The GIF doesn't capture it, but it is faster.

I also added the documentation file, your thoughts there are welcome.

**Comment by jasmussen** [comment]
The animation uses `translate` only, so it's really fast, and it seems to work well in the simulator also:

simulator

**Comment by jasmussen** [comment]
Aside from the the docs and additional feedback, it feels like the primary next step to figure out is _the point of origin_. It is partially encapsulated in these comments by @gziolo and @youknowriad:

> Maybe instead, this would be a HoC/renderprop/hook adding the correct className based on the "position" (top left,...)

> Ideally these styles shoudn't be added here and instead added to each respective component's stylesheet.

> Or generalized and controlled by the value passed through prop translated into a corresponding class name.

> If this animation is tied to the "popover" component, I guess this style should go there but if it's a very generic animation that can be applied anywhere, maybe we could try working on an "animatation" stylesheet loaded as a dependency where it's being used.

These all speak to the same point: this animation needs a point of origin. Which as noted in the docs is a key principle we have to consider — the animation from the ellipsis button and outwards can help create a sense of place and origin of the menu. 

There are a few challenges with that, though:

- These animations have to be RTL aware
- These animations have to be context aware. For example [as illustrated by Kjell](https://github.com/WordPress/gutenberg/pull/13617#issuecomment-459470157), sometimes the popover will upen upwards, and then the transform origin has to change to accommodate that. 

If, to @gziolo's suggestion, we could work the transform-origin property into the popover component itself, it seems like we could address all those issues, including CSS context feedback such as https://github.com/WordPress/gutenberg/pull/13617#discussion_r252648169. Can you help here Grzegorz?

**Comment by youknowriad** [comment]
I pushed an update here using a component. The idea is that we'd use this component for several kind of animations. 
I also applied the animation automatically to all popovers except the tooltips.
If you all agree with the approach, I think we should document the props of this component instead of documenting the classNames.

Let me know what you think @gziolo @jasmussen

**Comment by jasmussen** [comment]
This is amazing Riad, thanks so much!

animations

This appears solid to me. Very very nice. Notice how the document outline correctly animates from top center as it should. Solid!

I'm applying the Docs label because to me it seems like the next step is just going over the text a little bit.

**Comment by youknowriad** [comment]
I fixed the issue mentioned above, for me this is ready. Feel free to take over (README of the component...), if there's need for code tweaks, I'm here to help.

**Comment by jasmussen** [comment]
You do so fast and amazing work Riad, it's a gosh darn privilege to have your help on this.

**Comment by youknowriad** [comment]
I think this PR is very important and open the door to a lot of small animations (using this component) into other places of the UI.

**Review by kjellr — APPROVED** [review]
Awesome work, everyone! I pushed some super minor spelling/grammar updates to the new animation doc. 

From a design + documentation stand point, this looks great to me!

**Comment by jasmussen** [comment]
😍😍😍

Thank you everyone for the reviews! Let's ship this!

🤟

**Comment by youknowriad** [comment]
I think we should add a quick `README` for the `Animate` component before shipping.

**Comment by jasmussen** [comment]
Good point. I also want the tests to pass, I saw them failing earlier.

**Comment by youknowriad** [comment]
Mmm right, the tests are failing, like due to the very small delay for the links to be visible. I'll plan to look at this on monday though.

File: docs/designers-developers/designers/animation.md
Comment (chrisvanpatten) [inline docs/designers-developers/designers/animation.md]: Curious about the reference to animating on a curve, assuming this is referring to an easing curve (not literally a shape animation). I think that can often help an animation feel smoother… any particular "feel" this is intended to discourage?

File: packages/components/src/style.scss
Comment (chrisvanpatten) [inline packages/components/src/style.scss]: Is this a typo?

Comment (chrisvanpatten) [reply packages/components/src/style.scss]: (The double slash, to be clear!)

File: docs/designers-developers/designers/animation.md
Comment (jasmussen) [reply docs/designers-developers/designers/animation.md]: No no, this animation is a cubic bezier for the right easing. I was referring to the curve of the animation path. But the fact that this came up makes me think the language should not clarified 👍

Comment (chrisvanpatten) [reply docs/designers-developers/designers/animation.md]: @jasmussen Got it that makes a lot more sense :)

File: packages/components/src/animate/index.js
Comment (gziolo) [inline packages/components/src/animate/index.js]: This component should be documented on its own to ensure developers can use it to animate their own components. To be honest, I'm not quite sure how one could use it inside Gutenberg, too.

**Comment by gziolo** [comment]
> I think we should add a quick `README` for the `Animate` component before shipping.

I think this is the only blocker. Let me know if you need any help with this one. One great example to follow:
https://github.com/WordPress/gutenberg/tree/master/packages/components/src/dropdown-menu

You can leave JavaScript snippet to me once everything else is well documented.

File: docs/designers-developers/designers/animation.md
Comment (gziolo) [inline docs/designers-developers/designers/animation.md]: Ideally, we never promote those class names but rather convert them to components and keep them as an implementation detail.

Comment (jasmussen) [reply docs/designers-developers/designers/animation.md]: I strongly agree. In fact I feel that documenting these like this is the first step to retiring them in favor of something better/simpler.

Comment (gziolo) [reply docs/designers-developers/designers/animation.md]: Sounds like a plan :)

File: packages/components/src/animate/README.md
Comment (gziolo) [inline packages/components/src/animate/README.md]: @jasmussen - this is just template, it needs more work :)

**Comment by gziolo** [comment]
@youknowriad - did you investigate failing e2e tests?

**Comment by youknowriad** [comment]
I didn't have the time to look yet.

**Comment by adamsilverstein** [comment]
updated the docs manifest in https://github.com/WordPress/gutenberg/pull/13617/commits/e62c7397f77f339719310670953377109c75b4d2 which should fix failing tests

**Comment by jasmussen** [comment]
You are the greatest! Thank you Adam 💖

**Comment by adamsilverstein** [comment]
Tests are failing here because of the newly introduced delay between clicking the block toolbar, more menu & block library. We don't account for this in the e2e tests so after 'clicking' the opener element, we immediately check for the expected opened element which now fails.

We can introduce a brief delay in each test after performing the click action to fix this. Another option might be to disable animations for testing, however this seems more prone to introducing unintended side effects.

**Comment by aduth** [comment]
While I agree that adding differences between the test execution and user-facing behaviors is prone to bugs, we should be extremely vigilant against adding measurable delays to the runtime of the tests, as these will add seconds or minutes to run time of every build.

In this case, my biggest concern would be that the implementation of the `insertBlock` utility function [operates using the inserter](https://github.com/WordPress/gutenberg/blob/master/packages/e2e-test-utils/src/insert-block.js). This function is called at least 91 times over the course of the current set of tests, and will be expected to grow over time. If we are going to have delays, they should be as contained as possible. I'm personally fine with utility functions which "cheat" and operate by the most direct paths possible, since the utilities are not meant as substitutes for tests, and user interaction of the equivalent utility behavior should be tested in its own test cases.

In other words, if the inserter has a delay as a result of its opening animation, we should change the implementation of `insertBlock` utility to use something like the underlying data actions instead.

See prior art:

- https://github.com/WordPress/gutenberg/blob/master/packages/e2e-test-utils/src/select-block-by-client-id.js
- https://github.com/WordPress/gutenberg/blob/master/packages/e2e-test-utils/src/set-post-content.js
- https://github.com/WordPress/gutenberg/blob/609a42c/packages/e2e-test-utils/src/create-new-post.js#L31-L34

**Comment by adamsilverstein** [comment]
Digging a little further, the tests are failing after `transformBlockTo` calls, https://github.com/WordPress/gutenberg/blob/b786aad41b4670190b6cac4d99bab3abd3f85049/packages/e2e-test-utils/src/transform-block-to.js - this uses mouse movement to activate the transform menu.  Looks like that is called around 20 times in the test suite. I'll see if this is fixable by reimplementation rather than adding a delay.

**Comment by adamsilverstein** [comment]
@aduth any tips on how to transform the blocks directly, replacing the mouse manipulation in `transformBlockTo`? looking at the `BlockSwitcher` it is dispatching `( 'core/editor' ).replaceBlocks` which takes `ownProps.clientIds` - can we get that somewhere?

**Comment by aduth** [comment]
There might be a way to do it through actions. Could be simpler though; I think the original reason for the mouse moves may no longer be necessary, as we've since added support that an "Escape" keypress would activate the selected block's toolbar. As far as I understand, that was why we had some of these `page.mouse.move` in the E2E tests.

Related: #10906

**Comment by youknowriad** [comment]
@adamsilverstein Not sure about removing them but they can probably be replaced by a "Escape" key press.

**Comment by adamsilverstein** [comment]
> There might be a way to do it through actions

I haven't figured out how to fix each of these yet, in the meantime I tested with (110ms) delays when opening menus with animation - if only to figure out what needs adjusting - in https://github.com/adamsilverstein/gutenberg/pull/1

**Comment by youknowriad** [comment]
@adamsilverstein I'm trying to take a similar approach to your PR here to try to fix the tests.

The difference is I named the delay function, "waitForAnimation" to clarify that it's not something we should use everywhere but explicitly when there are animations ongoing. Using delays randomly can hide real issues in e2e tests.

**Comment by youknowriad** [comment]
I did some tweaks to the e2e tests, I'm still evaluating if there are other places where I missed the `waitForAnimation` call. I think it's not a perfect solution but it's an ok one for now. We shouldn't block the PR because we don't have a perfect solution for e2e tests. the important thing is that the e2e tests pass consistently.

On the docs. I pushed tweaks to the Animate README. I'd personally prefer if we remove the animations doc entirely or at least rewrite it to link to the README of the Animate component instead. I don't want these classnames to be documented, that way people won't start using them in plugins...

**Comment by jasmussen** [comment]
> I'd personally prefer if we remove the animations doc entirely or at least rewrite it to link to the README of the Animate component instead. 

Do it! The argument is solid and now that we have a component for it, we have a path forward to making the animations reusable in a better way.

File: docs/designers-developers/designers/animation.md
Comment (youknowriad) [inline docs/designers-developers/designers/animation.md]: @nosolosw @mkaz @chrisvanpatten not sure how to add this to the handbook/manifest with the new setup. Feel free to push a commit to add it.

**Review by youknowriad — APPROVED** [review]
LGTM 👍 

Another review would be good as I worked a lot on this PR.

**Comment by jasmussen** [comment]
You're the greatest. All of you.

File: packages/components/src/animate/README.md
Comment (gziolo) [inline packages/components/src/animate/README.md]: When `children` isn't a function it will explode at the moment.

File: packages/components/src/animate/index.js
Comment (gziolo) [inline packages/components/src/animate/index.js]: Should it exit early when children isn't a function?

**Review by gziolo — CHANGES_REQUESTED** [review]
I applied two small tweaks to docs myself.

There is one use case where the usage of `Animate` might be problematic.

File: packages/components/src/animate/README.md
Comment (youknowriad) [reply packages/components/src/animate/README.md]: I think it's fine, it's a requirement for the component. it's like saying the `options` is an object and if you pass a number it will explode.

I do see the specificity in JSX about the `children` prop but ultimately I think we can just consider it a requirement to be a function.

File: packages/components/src/animate/index.js
Comment (youknowriad) [reply packages/components/src/animate/index.js]: Same as I was saying above, we don't guard about the other props. I'm hesitant to add a special case for children.

File: docs/designers-developers/designers/animation.md
Comment (chrisvanpatten) [reply docs/designers-developers/designers/animation.md]: I don't think I'll be able to push a commit myself today :( But @mkaz can help I bet? The tl;dr is that you just need to add an entry in docs/toc.json and run the docs build and the manifest will be updated from that file.

File: packages/e2e-test-utils/src/wait-for-animation.js
Comment (aduth) [inline packages/e2e-test-utils/src/wait-for-animation.js]: Understanding the desire that this is introduced to help unblock the pull request, I hesitate on whether this should be considered an "unstable" API or not. I think as proposed it's at least doing well enough to not assume that it operates on some timer, since I think an implementation leveraging [`transitionend`](https://developer.mozilla.org/en-US/docs/Web/Events/transitionend) is probably a worthwhile exploration for further improvement. But the bigger issue to me is forcing it to be a conscious part of someone's mind in authoring tests, and whether that's strictly necessary. It seems far too prone to being forgotten and causing unexpected test failures, though if they're obvious enough and consistent failures, perhaps they're not truly "unexpected" and it's okay.

In short, I wonder if it would be a good idea to mark this as unstable, or perhaps not if we think that even if it be part of a public API (and usable directly) that we could find some other way to "automate" this on behalf of a test author in the future.

Comment (aduth) [reply packages/e2e-test-utils/src/wait-for-animation.js]: Other possibilities to avoid it being a conscious part of the developer test authoring workflow is to ensure that most of the interactions which involve an animation are sufficiently abstracted. So rather than a test including:

```js
await page.click( 'button[aria-label="More options"]' );
await waitForAnimation();
```

We'd see something like `await openMoreOptions();` which handles the animation implementation detail internally.

File: packages/components/src/animate/README.md
Comment (gziolo) [reply packages/components/src/animate/README.md]: I think it's a conceptual issue with render props :)

File: packages/e2e-test-utils/src/wait-for-animation.js
Comment (youknowriad) [reply packages/e2e-test-utils/src/wait-for-animation.js]: I do agree we should "automate" this on behalf of the test author behind utilities like:

 - insert block
 - click menu item
 - switch to editor mode
...

at that point we could hide it entirely from the public API

A lot of tests need to be updated though, and I don't think that's the PR for it.

**Review by gziolo — APPROVED** [review]
I get your explanations, it isn't perfect but this is what it is in every other place we use render props.

Feel free to merge when documentation is properly linked and suggestion from @aduth is addressed.

File: packages/e2e-test-utils/src/wait-for-animation.js
Comment (youknowriad) [reply packages/e2e-test-utils/src/wait-for-animation.js]: @aduth Looking at the tests, I don't think we can entirely abstract all the interactions without creating very specific functions. One example that comes to mind, is do we want to add an  `openFontSizePicker` utility where it's very specific to a given a block in a given state?

Comment (aduth) [reply packages/e2e-test-utils/src/wait-for-animation.js]: > A lot of tests need to be updated though, and I don't think that's the PR for it.

As long as there's some reasonably clear path forward (and with at least a nod to the fact we're adding technical debt) I'm fine to not let it block the work here. It would be good to have an issue describing the future planned work.

Comment (youknowriad) [reply packages/e2e-test-utils/src/wait-for-animation.js]: I honestly don't have a clear path forward on my mind. I'd love thoughts. 

Abstractions over the generic use-cases (block menus, global menus, inserter) are a good solution but what about the very specific use-cases?

Comment (aduth) [reply packages/e2e-test-utils/src/wait-for-animation.js]: > One example that comes to mind, is do we want to add an `openFontSizePicker` utility where it's very specific to a given a block in a given state?

No, probably not. But then it resurfaces the problem of imposing the need to be aware of a delay to the test author. My original intent with raising the point was in trying to see through the implications, and if not to solve it in place, have some idea for a future goal. To me, that future goal ideally wouldn't involve `waitForAnimation` existing, hence why I'd considered whether to mark it as unstable.

The alternatives I'd have in mind which bypass the need for the function are rough, and in many ways similarly flawed / technically challenging:

- Not have the transitions exist for the tests. It's still an issue that these are going to add measurable delays to the runtime of every build.
- In some way, every `transitionstart` which ever occurs on a page is monitored such that when a test includes an interaction (e.g. click) which either triggers / follows (unsure which yet) a transition, the following action would be delayed automatically.

I'm also open to saying that accepting that this needs to be consciously considered as part of writing an end-to-end tests. I'm sure there's an argument that these "steps" follow the user's actions, which includes that when a user clicks a button, they may have to wait for the animation.

tl;dr: I have concerns, and I wanted them to be noted, but I also think this is good enough in its current state to be able to move forward on.

Comment (youknowriad) [reply packages/e2e-test-utils/src/wait-for-animation.js]: Thanks for the thoughts, I'm personally not concerned about stable/unstable API since it's just a test utility that said, I'll follow up with an issue with these great thoughts.

I just had an idea probably similar to your first point here: maybe in the tests we load a specific stylesheet forcing all animations to be very quick (unnoticeable by tests) like 1ms. 

 - It ensures we're testing the existence of the animation (like prod)
 - It ensures the tests are stable.

Comment (aduth) [reply packages/e2e-test-utils/src/wait-for-animation.js]: > I'm personally not concerned about stable/unstable API since it's just a test utility

Speaking of breaking changes which would require a major version bump to `@wordpress/e2e-test-utils`: We should include a note (enhancement, minor bump) for the addition of `waitForAnimation` and ideally document it somewhere if we're doing that for these utilities (I'd suspect not).

Comment (youknowriad) [reply packages/e2e-test-utils/src/wait-for-animation.js]: Actually that package hasn't been released yet. It feels weird to add just `waitForAnimation` to the changelog where it only includes a single entry `Initial release`.

Good call though, I added the `Animate` component to the components changelog.

Comment (aduth) [reply packages/e2e-test-utils/src/wait-for-animation.js]: > It feels weird to add just `waitForAnimation` to the changelog where it only includes a single entry `Initial release`.

Oh, I didn't realize it hadn't been released yet. It's fine to just leave it then as "Initial release".

**Comment by youknowriad** [comment]
Let's animate this editor

File: docs/designers-developers/designers/animation.md
Comment (mkaz) [reply docs/designers-developers/designers/animation.md]: Looks like it was added, let me know if I can help

**Comment by aduth** [comment]
Related issue in use of `Animate` on the URL popover: #14872

File: packages/components/src/animate/README.md
Comment (aduth) [inline packages/components/src/animate/README.md]: Is there a reason that `options` ought to be a separate object, vs. just having its own properties be props of the component itself?

In other words, is...

```jsx
<Popover type="appear" options={ { origin: 'top' } } />
```

...necessarily any better than:

```jsx
<Popover type="appear" origin="top" />
```

?

Comment (youknowriad) [reply packages/components/src/animate/README.md]: My thinking at the time is that "options" are type-specific, their format change for each animation type.

Comment (aduth) [reply packages/components/src/animate/README.md]: > My thinking at the time is that "options" are type-specific, their format change for each animation type.

Okay, thanks for clarifying. I'm not sure it would necessarily be _wrong_ to still flatten these as top-level props, even if each `type` had their own handling and own set of options, but the rationale you give is sensible.

It may have been misleading to me because in the current implementation, the only option is `origin`, and it is consistent for both the `appear` and `slide-in` types, though I suppose that's more a reflection of its current state, and these may be expanded at some point in the future.

**Pull request merged**
